util/extsort/disk_sorter.go

```diff
@@ -131,29 +900,363 @@ func (d *DiskSorter) Sort(ctx context.Context) error {
+				files = append(files, file)
+				break
+			}
+		}
+	}
+	logger.Info(
+		"max overlap depth reached the compaction threshold, pick files to compact",
+		zap.Int("maxDepth", maxDepth),
+		zap.Int("threshold", compactionThreshold),
+		zap.Int("fileCount", len(files)),
+	)
+	return files
+}
+
+func splitCompactionFiles(files []*fileMetadata, maxCompactionDepth int) [][]*fileMetadata {
+	// Split files into non-overlapping groups.
+	slices.SortFunc(files, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	})
+	var groups [][]*fileMetadata
+	curGroup := []*fileMetadata{files[0]}
+	maxEndKey := files[0].endKey
+	for _, file := range files[1:] {
+		if bytes.Compare(file.startKey, maxEndKey) >= 0 {
+			groups = append(groups, curGroup)
+			curGroup = []*fileMetadata{file}
+		} else {
+			curGroup = append(curGroup, file)
+		}
+		if bytes.Compare(file.endKey, maxEndKey) > 0 {
+			maxEndKey = file.endKey
+		}
+	}
+	if len(curGroup) > 0 {
+		groups = append(groups, curGroup)
+	}
+
+	var finalGroups [][]*fileMetadata
+	// Compact each group of files.
+	for _, group := range groups {
+		numSubGroups := (len(group)-1)/maxCompactionDepth + 1
+		subGroupSize := (len(group)-1)/numSubGroups + 1
+		for i := 0; i < len(group); i += subGroupSize {
+			j := i + subGroupSize
+			if j > len(group) {
+				j = len(group)
+			}
+			finalGroups = append(finalGroups, group[i:j])
+		}
+	}
+	return finalGroups
+}
+
+type compaction struct {
+	startKey []byte
+	endKey   []byte
+	// overlapFiles are files that overlap with the compaction range.
+	// They are ordered by their start keys.
+	overlapFiles []*fileMetadata
+}
+
+func buildCompactions(files []*fileMetadata, maxCompactionSize int) []*compaction {
+	var startKey, endKey []byte
+	buckets := make([]kvStatsBucket, 0, len(files))
+	for _, file := range files {
+		buckets = append(buckets, file.kvStats.Histogram...)
+		if startKey == nil || bytes.Compare(file.startKey, startKey) < 0 {
+			startKey = file.startKey
+		}
+		if endKey == nil || bytes.Compare(file.endKey, endKey) > 0 {
+			endKey = file.endKey
+		}
+	}
+	// If there is kv stats, return a single compaction for all files.
+	if len(buckets) == 0 {
+		overlapFiles := slices.Clone(files)
+		slices.SortFunc(overlapFiles, func(a, b *fileMetadata) bool {
+			return bytes.Compare(a.startKey, b.startKey) < 0
+		})
+		return []*compaction{{
+			startKey:     startKey,
+			endKey:       endKey,
+			overlapFiles: overlapFiles,
+		}}
+	}
+
+	slices.SortFunc(buckets, func(a, b kvStatsBucket) bool {
+		return bytes.Compare(a.UpperBound, b.UpperBound) < 0
+	})
+	// Merge buckets with the same upper bound.
+	n := 0
+	for i := 1; i < len(buckets); i++ {
+		if bytes.Equal(buckets[n].UpperBound, buckets[i].UpperBound) {
+			buckets[n].Size += buckets[i].Size
+		} else {
+			n++
+			buckets[n] = buckets[i]
+		}
+	}
+	buckets = buckets[:n+1]
+
+	var (
+		kvSize      int
+		compactions []*compaction
+	)
+	for i, bucket := range buckets {
```

Comment:
seems we assume that for a bucket, the size are at the point of UpperBound, rather than distributed evenly in (previous UpperBound, UpperBound]? maybe left a comment to explain this estimation

Reply:
If data is uniformly distributed in each file, size being at the point is equivalent to size being distributed evenly in `(previous UpperBound, UpperBound]`.